File: internal/unlink/unlink.c
/*
 * © 2024 Snyk Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//go:build ignore

#include "common.h"
#include "bpf_tracing.h"

char __license[] SEC("license") = "Dual MIT/GPL";

struct trace_event_raw_sys_unlinkat {
	u64 _unused;
    u32 nr;
    u64 dfd;
    char *pathname;
    u64 flags;
    u64 mode;
};

struct trace_event_raw_sys_unlink {
	u64 _unused;
    u32 nr;
    char *pathname;
    u64 mode;
};

struct event {
	__u32 pid;
	__u8 pathname[255];
	__u64 timestamp;
};

const struct event *unusedevent __attribute__((unused));

struct {
	__uint(type, BPF_MAP_TYPE_RINGBUF);
	__uint(max_entries, 1 << 24);
} events SEC(".maps");

static __always_inline int unlink_helper(const char *pathname)
{
	struct event *event;
	event = bpf_ringbuf_reserve(&events, sizeof(struct event), 0);
	if (!event) return 0;

	u64 id = bpf_get_current_pid_tgid();
	u32 pid = id;
	event->pid = pid;
	u64 timestamp = bpf_ktime_get_ns();
	event->timestamp = timestamp;
	bpf_probe_read_user_str(&event->pathname, sizeof(event->pathname), pathname);
	bpf_ringbuf_submit(event, 0);
	return 0;
}

SEC("tracepoint/syscalls/sys_unlink")
int trace_unlink(struct trace_event_raw_sys_unlink *ctx)
{
	return unlink_helper(ctx->pathname);
}

SEC("tracepoint/syscalls/sys_unlinkat")
int trace_unlinkat(struct trace_event_raw_sys_unlinkat *ctx)
{
	return unlink_helper(ctx->pathname);
}

File: internal/unlink/unlink.go
/*
 * © 2024 Snyk Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package unlink

import (
	"ebpf-detector/internal/logger"
	"ebpf-detector/internal/utils"
	"errors"
	"runtime"
	"unsafe"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/perf"
	"github.com/cilium/ebpf/ringbuf"
)

//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target amd64 -type event bpf unlink.c -- -I../../headers
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target arm64 -type event bpf unlink.c -- -I../../headers

type Event struct {
	Pid             uint32
	Path            string
	TimestampMicros uint64
}

type ExploitDetectionHandler func(*Event)

func Listen(handler ExploitDetectionHandler) {
	objs := bpfObjects{}
	if err := loadBpfObjects(&objs, nil); err != nil {
		logger.Error(err, "Failed loading unlink bpf objects")
		return
	}
	defer objs.Close()

	tracepointSet := false
	if runtime.GOARCH != "arm64" {
		unlink, err := link.Tracepoint("syscalls", "sys_enter_unlink", objs.TraceUnlink, nil)
		if err != nil {
			logger.Error(err, "Failed setting unlink tracepoint")
		} else {
			tracepointSet = true
			defer unlink.Close()
		}
	}

	unlinkat, err := link.Tracepoint("syscalls", "sys_enter_unlinkat", objs.TraceUnlinkat, nil)
	if err != nil {
		logger.Error(err, "Failed setting unlinkat tracepoint")
	} else {
		tracepointSet = true
		defer unlinkat.Close()
	}

	if !tracepointSet {
		return
	}

	readOpenEvents(objs.Events, handler)
}

func readOpenEvents(events *ebpf.Map, handler ExploitDetectionHandler) {
	logger.Info("Listening on unlink events")
	reader, err := ringbuf.NewReader(events)
	if err != nil {
		logger.Error(err, "Failed reading unlink events map")
		return
	}

	for {
		record, err := reader.Read()
		if err != nil {
			if errors.Is(err, perf.ErrClosed) {
				return
			}

			return
		}

		bpfEvent := (*bpfEvent)(unsafe.Pointer(&record.RawSample[0]))
		pathname := utils.ConvertCString(bpfEvent.Pathname[:])
		timestampMicros := bpfEvent.Timestamp / 1000 // eBPF timestamp is in nanos
		handler(&Event{bpfEvent.Pid, pathname, timestampMicros})
	}
}
